File: mobile/src/api/transcriptions.ts
import { apiRequest } from './client';

/** Mirrors the backend's transcription state machine EXACTLY (see
 * backend/app/db/models/transcription.py) — this is a different state machine
 * from the mobile app's own SyncStatus ("did the server receive the audio?").
 * This answers "did AI processing turn the audio into text?" */
export type TranscriptionStatus = 'pending' | 'processing' | 'completed' | 'failed';

export interface TranscriptionResponse {
  id: string;
  submissionId: string;
  status: TranscriptionStatus;
  transcript: string | null;
  languageCode: string | null;
  errorMessage: string | null;
}

interface TranscriptionResponseWire {
  id: string;
  submission_id: string;
  status: string;
  transcript: string | null;
  language_code: string | null;
  error_message: string | null;
}

function fromWire(wire: TranscriptionResponseWire): TranscriptionResponse {
  return {
    id: wire.id,
    submissionId: wire.submission_id,
    status: wire.status as TranscriptionStatus,
    transcript: wire.transcript,
    languageCode: wire.language_code,
    errorMessage: wire.error_message,
  };
}

/**
 * POST /api/v1/submissions/{submissionId}/transcribe. Starts transcription if
 * none is running yet, or simply reports the current state (without a second
 * provider call) if one is already 'processing' or already 'completed' — see
 * backend/README.md. Always resolves with the current true state; never
 * fabricates a transcript.
 */
export async function triggerTranscription(submissionId: string): Promise<TranscriptionResponse> {
  const wire = await apiRequest<TranscriptionResponseWire>(
    `/api/v1/submissions/${submissionId}/transcribe`,
    { method: 'POST' }
  );
  return fromWire(wire);
}

File: backend/app/api/routes/transcriptions.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.core.config import settings
from app.db.session import get_db
from app.schemas.transcription import TranscriptionRead
from app.services import submissions as submission_service
from app.services import transcriptions as transcription_service

router = APIRouter(prefix="/api/v1/submissions/{submission_id}", tags=["transcriptions"])


@router.post("/transcribe", response_model=TranscriptionRead)
def trigger_transcription(submission_id: UUID, db: Session = Depends(get_db)):
    """Starts (or retries) transcription for a voice submission's uploaded
    audio, or reports the current state without calling Sarvam again if one is
    already in flight or already completed. Always returns 200 with the current
    TranscriptionRead — 'processing'/'failed' are legitimate, honest states of a
    successfully-handled request, not HTTP-level errors; only a precondition
    that blocks starting an attempt at all (submission missing/wrong type/no
    audio yet, or the server has no Sarvam key configured) is a 4xx/5xx.
    See services/transcriptions.py:start_transcription for the full idempotency
    and concurrency strategy."""
    if not settings.sarvam_api_key:
        raise HTTPException(
            status_code=503, detail="Transcription service is not configured on the server."
        )

    try:
        transcription, _outcome = transcription_service.start_transcription(db, submission_id)
    except LookupError:
        raise HTTPException(status_code=404, detail="Submission not found")
    except transcription_service.SubmissionNotAudioCapableError:
        raise HTTPException(
            status_code=400,
            detail="Transcription only applies to submissions that can carry audio "
            "(capture_type 'voice' or 'explore')",
        )
    except transcription_service.AudioNotUploadedError:
        raise HTTPException(
            status_code=400, detail="Audio has not been uploaded for this submission yet"
        )

    return transcription


@router.get("/transcription", response_model=TranscriptionRead)
def get_transcription(submission_id: UUID, db: Session = Depends(get_db)):
    """Reads the current transcription state/result — never triggers a Sarvam
    call. 404 if the submission doesn't exist, or if it does but transcription
    was never started (no audio uploaded yet)."""
    submission = submission_service.get_submission(db, submission_id)
    if submission is None:
        raise HTTPException(status_code=404, detail="Submission not found")

    transcription = transcription_service.get_transcription_by_submission_id(db, submission_id)
    if transcription is None:
        raise HTTPException(
            status_code=404, detail="No transcription exists yet for this submission"
        )
    return transcription
